### PR TITLE
refactor: payments as subscription and as micropayments for requests

### PR DESCRIPTION
**Pull request by kupermind**
- Payments as subscription and as micropayments for requests

### CODE REVIEW

File: contracts/BalanceTrackerFixedPriceBase.sol [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+error InvalidPayloadLength(uint256 provided, uint256 expected);
+
+/// @dev Failure of a transfer.
+/// @param token Address of a token.
+/// @param from Address `from`.
+/// @param to Address `to`.
+/// @param amount Amount value.
+error TransferFailed(address token, address from, address to, uint256 amount);
+
+abstract contract BalanceTrackerFixedPriceBase {
```
Comment (kupermind) [inline contracts/BalanceTrackerFixedPriceBase.sol:49]: The common ground of everything fixed price

File: contracts/BalanceTrackerFixedPriceBase.sol [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+        address requester,
+        uint256
+    ) external payable {
+        // Check for marketplace access
+        if (msg.sender != mechMarketplace) {
+            revert MarketplaceOnly(msg.sender, mechMarketplace);
+        }
+
+        // Check for native value
+        _checkNativeValue();
```
Comment (kupermind) [inline contracts/BalanceTrackerFixedPriceBase.sol:101]: This simply checks if working with a native token or another one.

File: contracts/BalanceTrackerFixedPriceBase.sol [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+        // Get account balance
+        uint256 balance = mapRequesterBalances[requester];
+
+        // Check the request delivery rate for a fixed price
+        if (balance < maxDeliveryRate) {
+            // Get balance difference
+            uint256 balanceDiff = maxDeliveryRate - balance;
+            // Adjust balance
+            balance += _getRequiredFunds(requester, balanceDiff);
+        }
```
Comment (kupermind) [inline contracts/BalanceTrackerFixedPriceBase.sol:115]: This is key - the balance can be top-upped via receive / deposit function. However if there's a lack of it - charge "at place", that makes the behavior exact equivalent to what there currently is.

File: contracts/BalanceTrackerFixedPriceBase.sol [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+        balance -= maxDeliveryRate;
+        mapRequesterBalances[requester] = balance;
+    }
+
+    /// @dev Finalizes mech delivery rate based on requested and actual ones.
+    /// @param mech Delivery mech address.
+    /// @param requester Requester address.
+    /// @param requestId Request Id.
+    /// @param maxDeliveryRate Requested max delivery rate.
+    function finalizeDeliveryRate(address mech, address requester, uint256 requestId, uint256 maxDeliveryRate) external {
```
Comment (kupermind) [inline contracts/BalanceTrackerFixedPriceBase.sol:127]: This is unchanged from previous version - except for the token field removal.

File: contracts/BalanceTrackerFixedPriceBase.sol [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+        // Record payment into mech balance
+        mapMechBalances[mech] += actualDeliveryRate;
+
+        emit MechPaymentCalculated(mech, requestId, actualDeliveryRate, rateDiff);
+    }
+
+    function _drain(uint256 amount) internal virtual;
+
+    /// @dev Drains collected fees by sending them to a Buy back burner contract.
+    function drain() external {
```
Comment (kupermind) [inline contracts/BalanceTrackerFixedPriceBase.sol:159]: Unchanged.

File: contracts/BalanceTrackerFixedPriceBase.sol [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+            revert ReentrancyGuard();
+        }
+        _locked = 2;
+
+        // Get mech balance
+        uint256 balance = mapMechBalances[msg.sender];
+        // TODO minimal balance value to account for the round-off
+        if (balance == 0 || balance < 10_000) {
+            revert InsufficientBalance(balance, 10_000);
+        }
```
Comment (kupermind) [inline contracts/BalanceTrackerFixedPriceBase.sol:196]: Need to define minimal value here to compare against.

Comment (kupermind) [reply contracts/BalanceTrackerFixedPriceBase.sol:196]: The rest is unchanged.

File: contracts/BalanceTrackerFixedPriceBase.sol [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+        mapMechBalances[msg.sender] = 0;
+
+        // Process withdraw
+        _withdraw(balance);
+
+        _locked = 1;
+    }
+
+    /// @dev Withdraws funds for a specific requester account.
+    function withdraw() external {
```
Comment (kupermind) [inline contracts/BalanceTrackerFixedPriceBase.sol:221]: Unchanged, just removed the token field

File: contracts/BalanceTrackerFixedPriceNative.sol [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+        if (_wrappedNativeToken == address(0)) {
+            revert ZeroAddress();
+        }
+
+        wrappedNativeToken = _wrappedNativeToken;
+    }
+
+    function _checkNativeValue() internal virtual override {}
+
+    function _getRequiredFunds(address, uint256 balanceDiff) internal virtual override returns (uint256) {
```
Comment (kupermind) [inline contracts/BalanceTrackerFixedPriceNative.sol:40]: Check for sufficient `msg.value` to top-up

File: contracts/BalanceTrackerFixedPriceToken.sol [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+    }
+
+    function _checkNativeValue() internal virtual override {
+        // Check for msg.value
+        if (msg.value > 0) {
+            revert NoDepositAllowed(msg.value);
+        }
+    }
+
+    function _getRequiredFunds(address requester, uint256 balanceDiff) internal virtual override returns (uint256) {
```
Comment (kupermind) [inline contracts/BalanceTrackerFixedPriceToken.sol:48]: Get top-up tokens

Comment (kupermind) [reply contracts/BalanceTrackerFixedPriceToken.sol:48]: probably needs to compare bool, right @77ph ?

File: contracts/MechFactoryFixedPriceNative.sol [hunk 10]
```diff
@@ -0,0 +1,46 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.28;
+
+import {MechFixedPriceNative} from "./MechFixedPriceNative.sol";
```
Comment (kupermind) [inline contracts/MechFactoryFixedPriceNative.sol:4]: Just the same as for the native token, no logic change

File: contracts/MechFixedPriceToken.sol [hunk 11]
```diff
@@ -0,0 +1,29 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.28;
+
+import {OlasMech} from "./OlasMech.sol";
+
+/// @title MechFixedPriceToken - Smart contract for OlasMech that accepts a fixed price payment for services in native token.
+contract MechFixedPriceToken is OlasMech {
```
Comment (kupermind) [inline contracts/MechFixedPriceToken.sol:7]: Same as a native token, just the payment type is different

File: contracts/MechMarketplace.sol [hunk 12]
```diff
@@ -351,8 +350,7 @@ contract MechMarketplace is IErrorsMarketplace {
         uint256 priorityMechServiceId,
         address requesterStakingInstance,
         uint256 requesterServiceId,
-        uint256 responseTimeout,
-        bytes memory paymentData
```
Comment (kupermind) [inline contracts/MechMarketplace.sol:355]: No need fir payment data, if anything additional is needed for the future - can be stored under `requestId` in the mech itself.

File: contracts/OlasMech.sol [hunk 13]
```diff
@@ -15,8 +15,9 @@ abstract contract OlasMech is Mech, IErrorsMech, ImmutableStorage {
     event RevokeRequest(address indexed sender, uint256 requestId);
 
     enum PaymentType {
-        FixedPrice,
-        Subscription
+        FixedPriceNative,
+        FixedPriceToken,
+        NvmSubscription
```
Comment (kupermind) [inline contracts/OlasMech.sol:20]: Giving better naming

File: contracts/OlasMech.sol [hunk 14]
```diff
@@ -214,6 +213,11 @@ abstract contract OlasMech is Mech, IErrorsMech, ImmutableStorage {
             return requestData;
         }
 
+        // Check for max delivery rate compared to requested one
+        if (maxDeliveryRate > mechDelivery.deliveryRate) {
+            revert Overflow(maxDeliveryRate, mechDelivery.deliveryRate);
+        }
```
Comment (kupermind) [inline contracts/OlasMech.sol:219]: This is important

File: contracts/MechMarketplace.sol (hunk 12 above)
Comment (DavidMinarsch) [reply contracts/MechMarketplace.sol:355]: Hm. I'd just leave a default `data` field.

File: contracts/BalanceTrackerFixedPriceBase.sol (hunk 6 above)
Comment (DavidMinarsch) [reply contracts/BalanceTrackerFixedPriceBase.sol:196]: My idea was to do the rounding up (ceil, not floor). This way, we don't need to define a hardcode on the balance as the marketplaceFee is always strictly positive (e.g. 1% fee, 1 cent withdrawn -> fee is 1 cent; 50 cent withdrawn -> fee is 1 cent, 1 dollar withdrawn -> fee is 1 cent

File: contracts/BalanceTrackerFixedPriceBase.sol [hunk 15]
```diff
@@ -0,0 +1,242 @@
...
+        _checkNativeValue();
+
+        // Get mech max delivery rate
+        uint256 maxDeliveryRate = IMech(mech).maxDeliveryRate();
+
+        // Get account balance
+        uint256 balance = mapRequesterBalances[requester];
+
+        // Check the request delivery rate for a fixed price
+        if (balance < maxDeliveryRate) {
```
Comment (DavidMinarsch) [inline contracts/BalanceTrackerFixedPriceBase.sol:110]: hm in the case of native value we don't account for native value being transferred in in all cases. this is a bug

Comment (kupermind) [reply contracts/BalanceTrackerFixedPriceBase.sol:110]: Good one

**Pull request merged**